File: nodes/csv-parser/csv-parser.process.js
export default async ({inputs, settings, config}) => {

  const csvString = inputs.csv || '';
  const delimiter = settings.delimiter || ',';
  const hasHeaders = settings.has_headers !== false;
  const trimValues = settings.trim_values !== false;
  const skipEmptyLines = settings.skip_empty_lines !== false;
  
  let data = [];
  let headers = [];
  let error = '';
  let success = false;
  
  try {
    // Split into lines
    let lines = csvString.split(/\r?\n/);
    
    // Skip empty lines if configured
    if (skipEmptyLines) {
      lines = lines.filter(line => line.trim() !== '');
    }
    
    if (lines.length === 0) {
      return { data: [], headers: [], error: '', success: true };
    }
    
    // Parse headers if configured
    if (hasHeaders && lines.length > 0) {
      headers = parseCSVLine(lines[0], delimiter, trimValues);
      lines = lines.slice(1);
    }
    
    // Parse data rows
    if (hasHeaders) {
      // Parse as objects with header keys
      data = lines.map(line => {
        const values = parseCSVLine(line, delimiter, trimValues);
        const row = {};
        
        // Map values to header keys
        headers.forEach((header, index) => {
          if (index < values.length) {
            row[header] = values[index];
          }
        });
        
        return row;
      });
    } else {
      // Parse as arrays of values
      data = lines.map(line => parseCSVLine(line, delimiter, trimValues));
    }
    
    success = true;
  } catch (e) {
    error = e.message || 'Failed to parse CSV';
    data = [];
    headers = [];
  }
  
  return { data, headers, error, success };
};

// Helper function to parse a CSV line respecting quotes
function parseCSVLine(line, delimiter, trim) {
  const values = [];
  let currentValue = '';
  let inQuotes = false;
  
  for (let i = 0; i < line.length; i++) {
    const char = line[i];
    const nextChar = i < line.length - 1 ? line[i + 1] : '';
    
    if (char === '"') {
      if (inQuotes && nextChar === '"') {
        // Escaped quote inside quotes
        currentValue += '"';
        i++; // Skip the next quote
      } else {
        // Toggle quote mode
        inQuotes = !inQuotes;
      }
    } else if (char === delimiter && !inQuotes) {
      // End of value
      values.push(trim ? currentValue.trim() : currentValue);
      currentValue = '';
    } else {
      // Add character to current value
      currentValue += char;
    }
  }
  
  // Add the last value
  values.push(trim ? currentValue.trim() : currentValue);
  
  return values;
} 

File: nodes/csv-parser/csv-parser.process.py
import csv
import io

async def process(inputs, settings, config, nodeConfig):
    """
    Process function for the CSV Parser node.
    Parses a CSV string into an array of objects or arrays.
    """
    csv_string = inputs.get("csv", "")
    delimiter = settings.get("delimiter", ",")
    has_headers = settings.get("has_headers", True)
    trim_values = settings.get("trim_values", True)
    skip_empty_lines = settings.get("skip_empty_lines", True)
    
    data = []
    headers = []
    error = ""
    success = False
    
    try:
        # Create a file-like object from the string
        csv_file = io.StringIO(csv_string)
        
        # Create CSV reader
        csv_reader = csv.reader(csv_file, delimiter=delimiter)
        
        # Read all rows
        rows = list(csv_reader)
        
        # Skip empty lines if configured
        if skip_empty_lines:
            rows = [row for row in rows if any(cell.strip() for cell in row)]
        
        if not rows:
            return {
                "data": [],
                "headers": [],
                "error": "",
                "success": True
            }
        
        # Parse headers if configured
        if has_headers and rows:
            headers = [h.strip() if trim_values else h for h in rows[0]]
            rows = rows[1:]
        
        # Parse data rows
        if has_headers:
            # Parse as objects with header keys
            for row in rows:
                row_dict = {}
                for i, value in enumerate(row):
                    if i < len(headers):
                        row_dict[headers[i]] = value.strip() if trim_values else value
                data.append(row_dict)
        else:
            # Parse as arrays of values
            data = [[cell.strip() if trim_values else cell for cell in row] for row in rows]
        
        success = True
    except Exception as e:
        error = str(e)
        data = []
        headers = []
    
    return {
        "data": data,
        "headers": headers,
        "error": error,
        "success": success
    } 